### DISCUSSION TITLE
Proof of Spacetime question

### DISCUSSION THREAD

**Discussion by anatoli26** [filecoin-project/rust-fil-proofs#1601] [Q&A]
Hi,

Not sure if it's the right place to ask this question, please advise.

I'm new to FileCoin, though I read most of the [docs.filecoin.io](https://docs.filecoin.io) and [proto.school/verifying-storage-on-filecoin](https://proto.school/verifying-storage-on-filecoin), but I can't find how exactly the PoST works.

My main doubt is about the merkle inclusion proofs on which PoST is apparently based. If I understand it correctly, you don't need to have the actual data, just the hashes, in order to complete the proof. If this is true, what prevents the storage provider to pre-calculate in advance all the needed hashes and then just drop the data?

If this is something trivial, please advice where to find the details to understand it better.

**Comment by porcuquine** [comment]
This is not a concern because the leaves of the merkle tree are 32-byte blocks of the replica itself. So the vanilla inclusion proofs actually contain the real (replicated) data, not only a hash of it. We did it that way specifically to prevent the issue you mention. I or someone else can point you to the best place to confirm this yourself in the code, but that's the short answer.

**Reply by anatoli26** [reply]
Thanks for your reply. So in this case the verifier should have the real data to perform the verification? Or it would just hash it and compare the result to the expected hash?

Maybe I haven't read some documentation, but I can't find the details of the exact mechanism of the PoST verification. Could you please point me to it? I can read the code, but if there's a higher-level explanation available of the entire PoST verification process, it would be much better.

**Reply by porcuquine** [reply]
> So in this case the verifier should have the real data to perform the verification?

It's not necessary, because the proof is a SNARK.

> Or it would just hash it and compare the result to the expected hash?

The merkle inclusion proofs are verified in a SNARK, so the PoSt verifier just has to verify the SNARK. The expected root is one of the public inputs, but all the internal hashes used in the inclusion proof — along with the leaf data — are private inputs. So the verifier doesn't need to have them.

> Maybe I haven't read some documentation, but I can't find the details of the exact mechanism of the PoST verification. Could you please point me to it?

Check the Filecoin Spec.
